contracts/Core.sol

```diff
@@ -92,11 +128,130 @@ contract Core is CoreInterface {
 		return coreChainIdRemote;
 	}
 
+	/// @dev public function openSTRemote
+	/// @return address coreOpenSTRemote
 	function openSTRemote()
 		public
 		view
 		returns (address /* OpenSTRemote */)
 	{
 		return coreOpenSTRemote;
 	}
+
+	/// @dev public function getLatestStateRootBlockHeight
+	/// @return uint256 latestStateRootBlockHeight
+	function getLatestStateRootBlockHeight()
+		public
+		view
+		returns (uint256 /* latest state root block height */)
+	{
+		return latestStateRootBlockHeight;
+	}
+
+	/// @dev public function getLatestStorageRootBlockHeight
+	/// @return uint256 latestStorageRootBlockHeight
+	function getLatestStorageRootBlockHeight()
+		public
+		view
+		returns (uint256 /* latest storage root block height */)
+	{
+		return latestStorageRootBlockHeight;
+	}
+
+	/**
+     * Commit new state root for a block height
+     *
+     */
+	/// @dev Commit new state root for a block height
+	/// @param _blockHeight block Number for which stateRoots mapping needs to update
+	/// @param _stateRoot state root of input block height
+	/// @return bytes32 stateRoot
+	function commitStateRoot(
+		uint256 _blockHeight,
+		bytes32 _stateRoot)
+		external
+		returns(bytes32 stateRoot)
+	{
+		// check if the caller is whitelisted worker
+		require(workers.isWorker(msg.sender), "Invalid worker address");
+		// State root should be valid
+		require(_stateRoot != bytes32(0), "Invalid state root");
+		// Input block height should be valid
+		require(_blockHeight > latestStateRootBlockHeight, "Given Block height is lower than latestBlockHeight.");
+
+		stateRoots[_blockHeight] = _stateRoot;
+		latestStateRootBlockHeight = _blockHeight;
+
+		emit StateRootCommitted(_blockHeight, _stateRoot);
+
+		return stateRoot;
+	}
+
+	/// @dev Verify account proof and update storageRoots mapping
+	/// @param _blockHeight block Number at which stake happened
+	/// @param _value rlpencoded => hashed account node object
+	/// @param _rlpParentNodes RLP encoded value of account proof parent nodes
+	/// @param _storageRoot storage root received from account proof
+    /// @return bool status
+	function proveOpenST(
+		uint256 _blockHeight,
+		bytes32 _value,
+		bytes _rlpParentNodes,
+		bytes32 _storageRoot)
+		external
+		returns(bool status)
+	{
+		// check if the caller is whitelisted worker
+		require(workers.isWorker(msg.sender), "Invalid worker address");
+		// Check for block height
+		require(_blockHeight != 0, "Invalid block height");
+		// Storage root should be valid
+		require(_storageRoot != bytes32(0), "Invalid storage root");
+
+		bytes32 stateRoot = stateRoots[_blockHeight];
+		// State root should be present for the block height
+		require(stateRoot != bytes32(0), "State root missing for given block height");
+
+		require(_value != bytes32(0), "Node values are missing");
+
+		// Encode remote address. Storage added because it was generating warning.
+		bytes memory encodedAddress = bytes32ToBytes(keccak256(coreOpenSTRemote));
```

Comment:
this is a constant value (before protocol upgrade in v0.9.4); let's calculate in constructor and not every time here